### PR TITLE
New GraphQLWsLink for graphql-ws subscriptions library

### LINKED ISSUE #8345: Why use WebSocketLink  from '@apollo/client/link/ws'
1) Sorry for my english.
2) thanks for your time.

In your documentation (https://www.apollographql.com/docs/react/api/link/apollo-link-ws/) you encourage to use WebSocket which depends on subscriptions-transport-ws npm package. but apollographql/subscriptions-transport-ws repo says 

> "The subscriptions-transport-ws library is not being actively maintained. It is recommended that you use the graphql-ws library instead."

Same case here https://www.apollographql.com/docs/graphql-subscriptions/authentication/ .

By other side in the next link you implement difference approach to the same solution, will be great or unify then, or explain pros/cons, or when use what:

https://www.apollographql.com/docs/graphql-subscriptions/express/
https://www.apollographql.com/docs/apollo-server/data/subscriptions/#example-authentication-with-onconnect

Finally I would like to ask you about an example to how implement graphql-ws with apollo server.

### PR DESCRIPTION
**Pull request by glasser**
Apollo Client currently contains `WebSocketLink` in
`@apollo/client/link/ws` which uses the `subscriptions-transport-ws`
library. That library is no longer actively maintained, and there is an
improved fork called `graphql-ws`.

The two libraries use different protocols so a different client link is
required for `graphql-ws`. (While the WebSocket protocol does allow for
subprotocol negotiation, neither server implementation supports this in
a practical way.)

This PR adds a new link `GraphQLWsLink` in
`@apollo/client/link/subscriptions`. Its constructor arguments are the
same as the `createClient` function in `graphql-ws` (or it can take a
`Client` object returned from that function), and you need to install
the optional peer dep `graphql-ws` instead of
`subscriptions-transport-ws`. Once you've created the link, it works
exactly like the old `WebSocketLink`.

This PR changes the main subscriptions doc page to mostly document the
new link, with an extra section at the bottom for the old link.

The core GraphQLWsLink code is based on MIT-licensed code from the
README of the graphql-ws repository.

Fixes #8345.
Part of https://github.com/apollographql/apollo-server/issues/6058

### CODE REVIEW

File: docs/source/api/link/apollo-link-subscriptions.md [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+```
+
+> **Note**: This link works with the newer `graphql-ws` library. If your server uses the older `subscriptions-transport-ws`, you should use the [`WebSocketLink` link from `@apollo/client/link/ws](./apollo-link-ws) instead.
+
+## Constructor
+
+```js
+import { GraphQLWsLink } from "@apollo/client/link/subscriptions";
+
+const link = new WebSocketLink({
```
Comment (lennyburdette) [inline docs/source/api/link/apollo-link-subscriptions.md:25]: Drive-by review: should this be GraphQLWsLink?

Comment (glasser) [reply docs/source/api/link/apollo-link-subscriptions.md]: Thanks! Fixed.

**Comment by glasser** [comment]
(fixed accidental prettier applications)

File: src/link/subscriptions/__tests__/graphqlWsLink.ts [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+function mockClient(subscribe: Client["subscribe"]): Client {
+  return {
+    subscribe,
+    // GraphQLWsLink doesn't use these methods
+    on: () => () => {},
+    dispose: () => {},
+  };
+}
+
+async function observableToArray<T>(o: Observable<T>): Promise<T[]> {
```
Comment (brainkim) [inline src/link/subscriptions/__tests__/graphqlWsLink.ts:42]: The tasteful programmer in me is saying let this lie, but the consistent programmer wants to see it moved to the utilities folder somewhere. I’m happy to see this remain here though.

File: src/link/subscriptions/__tests__/graphqlWsLink.ts [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+  return out;
+}
+
+describe("GraphQLWSlink", () => {
+  it("constructs", () => {
+    const client = mockClient(() => () => {});
+    expect(() => new GraphQLWsLink(client)).not.toThrow();
+  });
+
+  // TODO some sort of dependency injection
```
Comment (brainkim) [inline src/link/subscriptions/__tests__/graphqlWsLink.ts:54]: I don’t like the idea of dependency injection because it might cause impossible program states, but if we’re going to be creating GraphQL WS clients on behalf of the user when options are passed in, we should probably expose the created client on the link as a property (`link.client`).

File: src/link/subscriptions/index.ts [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+import { print, GraphQLError } from "graphql";
+import { createClient, ClientOptions, Client } from "graphql-ws";
+
+import { ApolloLink, Operation, FetchResult } from "../core";
+import { Observable } from "../../utilities";
+
+export class GraphQLWsLink extends ApolloLink {
+  private client: Client;
+
+  constructor(optionsOrClient: ClientOptions | Client) {
```
Comment (brainkim) [inline src/link/subscriptions/index.ts:37]: There is very little benefit to this sort of polymorphism, except maybe consistency and users not having to import `graphql-ws` directly. However, I think the latter pro can also be considered a a con, as people are likely to improperly add `graphql-ws` to their dependencies, or improperly configure the client and then open up an issue with us. Also, the type-sniffing of a dependency creeps me out because this would break if `graphql-ws` ever added a `subscribe` property directly to the client object. Therefore, I think it might be better to BYOWSC (bring your own websocket client).

But it’s not that big a deal.

**Review by brainkim — COMMENTED** [review]
Thanks for doing this @glasser. Some thoughts, but nothing blocking.

File: src/link/subscriptions/__tests__/graphqlWsLink.ts (hunk 3 above)
Comment (glasser) [reply src/link/subscriptions/__tests__/graphqlWsLink.ts:54]: This was copied from the existing ws link test. Happy to delete.

File: src/link/subscriptions/index.ts (hunk 4 above)
Comment (glasser) [reply src/link/subscriptions/index.ts:37]: That's my general philosophy as well. I was echoing the existing link here, but very happy to just require you to use graphql-ws directly. It simplifies the doc too. @benjamn Any thoughts or shall I make this change?

File: src/link/subscriptions/index.ts [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+                // reason will be available on clean closes
+                new Error(
+                  `Socket closed with event ${err.code} ${err.reason || ""}`
+                )
+              );
+            }
+
+            return sink.error(
+              new Error(
+                (err as GraphQLError[]).map(({ message }) => message).join(", ")
```
[the remark this answers is not in the document]
Comment (glasser) [reply src/link/subscriptions/index.ts:69]: This was copied from the graphql-ws README without much thought; happy to take whatever approach you deem best.

File: src/link/subscriptions/index.ts [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+
+  public request(operation: Operation): Observable<FetchResult> {
+    return new Observable((sink) => {
+      return this.client.subscribe<FetchResult>(
+        { ...operation, query: print(operation.query) },
+        {
+          next: sink.next.bind(sink),
+          complete: sink.complete.bind(sink),
+          error: (err) => {
+            if (err instanceof Error) {
```
[the remark this answers is not in the document]
Comment (glasser) [reply src/link/subscriptions/index.ts:54]: This was copied from the graphql-ws README without much thought; happy to take whatever approach you deem best.

File: src/link/subscriptions/__tests__/graphqlWsLink.ts (hunk 2 above)
Comment (glasser) [reply src/link/subscriptions/__tests__/graphqlWsLink.ts:42]: Up to you. I could put it in `src/utilities/observables/` but I wonder if that's the right place for something only used in a test.

File: src/link/subscriptions/index.ts [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+    super();
+    this.client =
+      // Client is an interface, not a class, so we can't use instanceof here
+      "subscribe" in optionsOrClient
+        ? optionsOrClient
+        : createClient(optionsOrClient);
+  }
+
+  public request(operation: Operation): Observable<FetchResult> {
+    return new Observable((sink) => {
```
[the remark this answers is not in the document]
Comment (glasser) [reply src/link/subscriptions/index.ts:47]: This was copied from the graphql-ws README without much thought; will do.

File: src/link/subscriptions/index.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (brainkim) [reply src/link/subscriptions/index.ts:54]: My thoughts are do this until people complain and then fix it if necessary.

File: src/link/subscriptions/__tests__/graphqlWsLink.ts (hunk 2 above)
Comment (brainkim) [reply src/link/subscriptions/__tests__/graphqlWsLink.ts:42]: Agreed!

File: src/link/subscriptions/index.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (benjamn) [reply src/link/subscriptions/index.ts:69]: Yikes, I see no reason this code supports the type assertion `err as GraphQLError[]`, which is immediately put to the test by calling the `.map` method.

Try this:
```ts
// May need to use a relative module specifier like "../../errors"
import { ApolloError } from "@apollo/client/errors"

sink.error(new ApolloError({
  graphQLErrors: Array.isArray(err) ? err : [err]
}))
```

**Review by benjamn — COMMENTED** [review]
Looking good, with a few questions/comments/responses. Thanks especially for all this documentation @glasser!

File: docs/source/api/link/apollo-link-ws.md [hunk 8]
```diff
@@ -1,7 +1,7 @@
 ---
 title: WebSocket Link
-sidebar_title: WebSocket
-description: Execute subscriptions (or other GraphQL operations) over WebSocket
+sidebar_title: WebSocket (newer protocol)
```
[the remark this answers is not in the document]
Comment (glasser) [reply docs/source/api/link/apollo-link-ws.md]: Oops!

File: src/link/subscriptions/__tests__/graphqlWsLink.ts (hunk 3 above)
Comment (glasser) [reply src/link/subscriptions/__tests__/graphqlWsLink.ts:54]: To clarify: if I'm changing to requiring you to pass in a client rather than client creation options, do you think it's still helpful to have this `link.client` field?

File: src/link/subscriptions/index.ts (hunk 4 above)
Comment (glasser) [reply src/link/subscriptions/index.ts:37]: OK, I think I want to move towards just taking the client (not creation options) and letting you import it yourself from `graphql-ws` rather than re-exporting it. I don't think importing a function from a package you're required to install yourself anyway (optional peer dep) is that big a deal.

At this point, the only non-test dependency on `graphql-ws` will be for the `Client` type (which is an exposed part of the GraphQLWsLink API).  Should we still have this as an optional peer dep?  Or is there a simpler alternative?  It would be nice if there was a mechanism in npm/ts for "type-only dep" but I don't think there really is one (devDependencies aren't appropriate for types that are actually used in the depending package's API, I think).  I guess we could also just copy the Client interface into the file or something but that's not great either.

File: src/link/subscriptions/index.ts [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+        { ...operation, query: print(operation.query) },
+        {
+          next: sink.next.bind(sink),
+          complete: sink.complete.bind(sink),
+          error: (err) => {
+            if (err instanceof Error) {
+              return sink.error(err);
+            }
+
+            if (err instanceof CloseEvent) {
```
[the remark this answers is not in the document]
Comment (glasser) [reply src/link/subscriptions/index.ts:58]: Hmm. I do think it would be nice to be able to use this link from Node (from tests, etc) so I think I'd like to copy over isLikeCloseEvent.

File: src/link/subscriptions/index.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (glasser) [reply src/link/subscriptions/index.ts:69]: That seems reasonable.

That said, the doc string on the `Sink.error` (in graphql-ws) this is documenting does say:

```
   * An error that has occured. Calling this function "closes" the sink.
   * Besides the errors being `Error` and `readonly GraphQLError[]`, it
   * can also be a `CloseEvent`, but to avoid bundling DOM typings because
   * the client can run in Node env too, you should assert the close event
   * type during implementation.
```

Which suggests that they are correct in assuming that these are the only three possibilities. That said I think your code here is good?

File: src/link/subscriptions/__tests__/graphqlWsLink.ts (hunk 2 above)
Comment (glasser) [reply src/link/subscriptions/__tests__/graphqlWsLink.ts:42]: I'm interpreting this as "not useful to put something test-only in that directory" so I'm doing nothing.

File: src/link/subscriptions/index.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (glasser) [reply src/link/subscriptions/index.ts]: bd92969

File: src/link/subscriptions/index.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (glasser) [reply src/link/subscriptions/index.ts]: 5eaa88c

File: src/link/subscriptions/__tests__/graphqlWsLink.ts (hunk 3 above)
Comment (glasser) [reply src/link/subscriptions/__tests__/graphqlWsLink.ts:54]: 942eec0

File: src/link/subscriptions/index.ts (hunk 4 above)
Comment (glasser) [reply src/link/subscriptions/index.ts]: 942eec0

File: docs/source/data/subscriptions.mdx [hunk 10]
```diff
@@ -312,3 +310,32 @@ The `useSubscription` Hook accepts the following options:
...
+
+## The older `subscriptions-transport-ws` library
+
+If your server uses `subscriptions-transport-ws` instead of the newer `graphql-ws` library, you need to make a few changes to how you set up your link.
+
+Instead of `npm install graphql-ws`, you `npm install subscriptions-transport-ws`
+
+Instead of `import { GraphQLWsLink } from '@apollo/client/link/subscriptions'`, you `import { WebSocketLink } from '@apollo/client/link/ws`.
+
+The options passed to the link constructor are slightly different. The subscriptions URL is specified in an `uri` option instead of an `url` option. The `connectionParams` option is nested under an options object called `options` instead of being at the top level.
```
[the remark this answers is not in the document]
Comment (glasser) [reply docs/source/data/subscriptions.mdx]: 2c03892

 I also changed the approach to documenting the old link to prefer constructing your own SubscriptionClient and passing it in, for parallelism with the only allowed approach for the new link.

File: src/link/subscriptions/index.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (glasser) [reply src/link/subscriptions/index.ts]: 464c9c6

File: src/link/subscriptions/index.ts (hunk 4 above)
Comment (benjamn) [reply src/link/subscriptions/index.ts]: I think the optional peer dep is still the way to go. Better to reuse the original `Client` type than copy it over.

File: docs/source/data/subscriptions.mdx (hunk 10 above)
[the remark this answers is not in the document]
Comment (benjamn) [reply docs/source/data/subscriptions.mdx]: As long as we keep the existing options API working in the `WebSocketLink` code and types, I’m fine with this documentation realignment.

**Comment by glasser** [comment]
@brainkim @benjamn I believe all feedback has been addressed.

File: docs/source/data/subscriptions.mdx (hunk 10 above)
[the remark this answers is not in the document]
Comment (glasser) [reply docs/source/data/subscriptions.mdx]: Yep, no code changes to `@apollo/client/link/ws` (I did fix one test that was copy-and-pasted from the test above it without the necessary change to test the right thing).

**Review by benjamn — APPROVED** [review]

**Comment by glasser** [comment]
@benjamn Rebased and retargeted to release-3.6, with CHANGELOG added.

**Review by brainkim — APPROVED** [review]
Thank you for bearing with my nits!

**Comment by darkbasic** [comment]
Would it be possible to release a new beta with this?
